scipy/_lib/_bunch.py

```diff
@@ -0,0 +1,102 @@
+
+import sys as _sys
+from keyword import iskeyword as _iskeyword
+
+
+class _TupleBunch:
+
+    _fields = ()
+    _extra_fields = ()
+
+    def __init__(self, **kwds):
+        # _validate_kwds isn't strictly necessary, since the code that follows
+        # it will raise an exception if the names in kwds are not valid, or if
+        # there is a name missing from kwds.
+        # The function provides a friendlier error message.
+        self._validate_kwds(kwds)
+        self.__dict__['_tuple'] = tuple(kwds[name] for name in self._fields)
+        for name in self._fields + self._extra_fields:
+            self.__dict__[name] = kwds[name]
+
+    def __repr__(self):
+        s = (self.__class__.__name__ + '(' +
+             ', '.join(f'{name}={self.__dict__[name]!r}'
+                       for name in self._fields + self._extra_fields) + ')')
+        return s
+
+    def __setattr__(self, name, value):
+        # Ensure that all attributes are read-only, and that new attributes
+        # can't be added to the object.
+        raise AttributeError("can't set attribute")
+
+    def _validate_kwds(self, kwds):
+        all_fields = self._fields + self._extra_fields
+        for name in kwds:
+            if name not in all_fields:
+                raise TypeError(f"unexpected keyword argument '{name}'")
+        for name in all_fields:
+            if name not in kwds:
+                raise TypeError(f"missing keyword argument '{name}'")
+
+    # Delegate a minimal set of dunder methods to self._tuple
```

Comment:
If you also add a `__iter__` method like:
```
    def __iter__(self):
        for ii in range(len(self)):
            yield self[ii]
```
You can inherit from `collections.abc.Sequence`:

https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes

Then you get some mixin methods for free, and people can iterate over the output (which they might have done previously) with the tuple.

Reply:
While I'm not saying we shouldn't do that, it looks like it isn't necessary--you can already iterate over the output:
```
In [7]: Result = make_tuple_bunch('Result', ['x', 'y'], ['w', 'beta'])

In [8]: result = Result(x=1, y=2, w=99, beta=0.5)

In [9]: for val in result:
   ...:     print(val) 
   ...:                                                                         
1
2
```